### ISSUE TITLE
RFC 302: Non-interactive protocols should not be in AIP v1.0

### ISSUE THREAD

**Issue by esplinr** [hyperledger/aries-rfcs#358]
We have concerns with the non-interactive protocols included in Aries Interop Profile v. 1.0. They could be used to undermine a user's privacy, and so they need more work before they can be included in an interopability profile.

Most protocols in the interop profile require user interaction to transmit data, but a few could trigger a response without any user interaction: 
* Trust Ping
* Feature Discovery
* Problem Report

These are important dev tools which we need in our ecosystem, but automatically responding to these requests could compromise the privacy of end users. We need to put more thought into the implications of these protocols on user interaction.

Some examples of the type of thinking these protocols still need:
* How often does an agent need to respond to these requests?
* How should a user be informed about these responses?
* Can a mediator like a cloud agent act on behalf of an edge agent to respond to these requests?

We would like to remove these protocols from the profile, and we will drive separate discussions for improving the RFCs that define these protocols.

**Comment by swcurran** [comment]
AIP 1.0.0 is to be used in the next few months and be replaced by new AIP versions referencing improved protocols as they are defined. I agree that unfettered use of those protocols could be a problem in the long term. However, their use in the short term is not (in my opinion) a significant risk, particularly in the limited use cases that will be deployed in the next few months. Mind you, if we don't get AIP defined and implemented, all PoCs will only be with pre-defined components vs. arbitrary implementations.

I suggest that your team work on RFCs to improve them to address the privacy concerns.  That is a separate issue from inclusion or exclusion from AIP 1.0.0.

Other thoughts:

- there is nothing to prevent an interaction based on the receipt of those requests. If you you are worried - ask the user before responding.
- While I think I can guess why Problem Report is included (responding to an unsolicited bad request - for which you could create an interaction), the inability to use Problem Report in valid situations will likely be, well, a problem.
- Feature discovery can return whatever you want, including nothing, or maybe just "Feature Discovery".  That's in the protocol today.  However in the early days of agents, it will likely be helpful for interoperability, even if we remove it in the future.

I vote we do not remove those from AIP 1.0.0.

**Comment by esplinr** [comment]
Responses:
* AIP 1.0.0 is meaningless if it is only appropriate for "limited use cases". Aries verifiable credentials are already being used in important production scenarios.
* AIP 1.0.0 will be an important milestone for much longer than a few months. If it is adopted in the real world (as we intend), we will need to maintain compatibility with the profile for some time to come.
* We will work on the RFCs to improve them to address privacy concerns, but we will not be adopting AIP 1.0.0 if they are included in their current form.
* Asking a non-technical user if they want to approve a "Trust Ping" is not a good idea. Let alone a "Feature Discovery" request.
* We don't feel as strongly about Problem Report, but included it for completeness.
* We agree that all of these protocols are important in development scenarios, but they need to be refined before *requiring* support for them in agents.
* Nothing prevents agents from implementing any of the Aries RFCs not included in the interoperability profile. We just can't depend on the existence of those RFCs in all compatible agents.

**Comment by swcurran** [comment]
Responses back :-)

- By "limited use case" I mean that in practice, we will only have a limited number of use cases (and users) in the next six months. Those cases will be real, and will only be limited by what business are willing to try. That in turns means that we don't have to be absolutely perfect.
- AIP 1.0.0 should not last longer than a few months. Or at least we hope it does not.  We want to see improvements to the protocols, new protocols to be added, and the community update mechanism to be used to enable continuous interop.
- It will be unfortunate if there is a split in the community on this because one member overrules the others.  But we need others to weigh in.
- I agree about a user being notified about such requests, and I wouldn't necessarily take that approach. However, if an app maker is particularly concerned that in the next six months, users will be harmed by unsolicited messages, that becomes an option. As does silently not responding or responding in a limited fashion (in the case of Feature Discovery).

Again, I strongly support providing input to the RFCs that (I think) should be helpful in defining AIP 1.1.0.

Thanks

**Comment by esplinr** [comment]
I don't understand the concern about "a split in the community on this because one member overrules the others". Exercises like this issue are how we achieve consensus together.

These RFCs are important for developer use cases, but I don't think they are required for credential exchange in the wild, which is the stated target of this Interop Profile. I think the quickest path to seeing that Interop Profile adopted and implemented is to keep the scope narrow. But I agree that an alternative approach is to wait on the Interop Profile until we are happy with the contents of the included RFCs.

**Comment by kdenhartog** [comment]
I don't have much of an opinion from an implementer's perspective, but I'll toss my NZD $.2 in from an attacker perspective, so take it with as much value as the currency exchange rate allows 😄

With regards to the problem report, I find this one to be the most problematic because it's likely to be the one that reveals the most information that allows an attacker to blackbox attack the agent. In my experience penetration testing, this was always extremely valuable insights to get if we could and it was usually an indication that someone left debugging mode on. The main reason we saw this as insightful was so we could easily identify vulnerabilities in unpatched software. 

Now I understand the value in providing useful problem reports and am not requesting that we remove them completely. Rather, I think there should be a more standardised approach to the types of errors that are returned in order to handle both the privacy concerns and reconnaissance concerns. In my opinion, this should be moved to something like HTTP codes where there's a standard set of return errors possible. Then if additional information is necessary in debugging mode, it can be returned as additional data inside the error.

**Comment by dhh1128** [comment]
@swcurran: I think the mental divide here isn't about one vendor or another; it's about the enterprise mindset (where automation is king) vs. the consumer mindset (where UX is king). EV has no problem responding to trust pings and feature discovery in an enterprise agent. Prompting app users who know nothing about the underlying plumbing/context is not good UX, plain and simple. Neither is automatically responding (because it's potentially unsafe). Knowing what to do for consumer apps is just simply a much harder problem for a consumer tool than it is in the enterprise.

@Richard: a way around this awkwardness in AIP's handling of consumer concerns might be to implement only the pinger (not pingee) role for trust ping, and only the requester (not responder) role for feature discovery, in Connect.Me. This would allow C.M to check some boxes without solving the much harder problem of how to confer with a user about correct behavior. Or we could implement the rest of the AIP and leave these unimplemented as footnoted exceptions.

**Comment by esplinr** [comment]
Discussed during Aries WG B on 2020-01-08.

We agreed to exclude from AIP 1.0 Trust Ping and Feature Discovery, because they are not critical for interoperability. But we will include Problem Report even with the potential privacy concerns because it is so important for validating that we have meaningful interoperability. Agents will need to decide how best to provide reports in a way that best protects their users.

**Comment by swcurran** [comment]
A point on the privacy concerns. It needs to be understood that there are only two ways for these messages to get to the user:

1. An entity that has established a connection to the recipient can send these unsolicited messages.
2. An out of band message such as we use for invitations or ephemeral challenges via either URLs or QR codes. I would argue that any mobile agent that silently handles such messages is not worth using.

For the first case, I think it would actually be very easy for a mobile app to provide a configuration and a notification of potential abuse - including an indicator of the abusing service.

In thinking about that I'm even more against dropping the protocols from the AIP.  I don't see any reason to drop them, since the security/privacy risk is both unlikely, configurable and easily detectable.

But I can live with it...

**Comment by swcurran** [comment]
While I hesitate to suggest that we get more nuanced on AIP I see one way we might do that in the future.  We use the enumerate list of agent types (the RFC lists 5 types), add a column for each to the RFC table, and add an indicator per type of agent that doesn't fully support a specific protocol. This would solve this issue.

I don't think that we should do that yet (extra complexity), but something to keep in mind if we have more challenges with this.

If we do this, we'd have to agree on a very constrained set of agent types (ideally, less than 5) and be ignore edge cases.

**Comment by dhh1128** [comment]
>there are only two ways for these messages to get to the user:

Way 3: an agency can forward an anoncrypted message that contains a `~service` decorator. That's not out-of-band, and it's also not using an existing connection. This is exactly how I would send spam in DIDComm land, if I got hold of someone's private DID.

>I think it would actually be very easy for a mobile app to provide a configuration and a notification of potential abuse - including an indicator of the abusing service.

I agree that It's easy to write one with a lousy UX (where config = all or nothing). Not so easy to detect abuse patterns and alert the user only when they "ought" to be interested. And very hard to explain the dynamics to a naive user so they can make intelligent config decisions.

> any mobile agent that silently handles such messages is not worth using :-).

There's a difference between decisions to silently ignore (which a mobile agent worth its salt *might* do) and decisions to silently cooperate (which no safe mobile agent should be doing). Maybe you were only alluding to the second one.

>the security/privacy risk is both unlikely, configurable and easily detectable.

**unlikely**: I agree today, but disagree in the long run (ping existed for a long time before network admins started turning it off as a source of abuse)

**configurable**: I disagree that this is trivially easy. What makes it hard is not the code you write, but how you explain to a naive user what they're configuring in a way that helps them make a good decision. Writing a dialog that asks users whether they want to accept unrecognized certs in a browser was easy; trying to unwind the consequences from millions of bad decisions has given browser manufacturers fits for a decade.

**easily detectable**: given that a single message could constitute the very privacy issue we're talking about (it doesn't have to be a recurring pattern), I don't see how "easily detectable" is true.

Anyway, I can live with the compromise, too, so we can make progress where we agree, and keep debating these issues elsewhere.

**Comment by swcurran** [comment]
The Way 3 method assumes that some sort of breach has already occurred - a bad actor has a private DID that they shouldn't have.  Even then, I think that can be mitigated with specific handling of such a message.  For example, ignoring the ~service decorator if you have a connection already.  As well, see below.

As to the lousy UX. Yes, it might be lousy for now, but if you can build a perfect UI up front, you are either a unicorn or moving too slowly.  Start with a "lousy" UX that is sufficient for now (have a default you can live with, and an explanation that is best you can do) and move forward. As well, see below.

Easily detectable. I would disagree with your level of concern over what is today very much an edge case (a malicious user exploiting this). As well, see below.

The below :-).  UX and mobile agents are iterated over continually. It won't be perfect today (or even tomorrow), it just has to be easily updateable to become continually better.  This gets back to the debate of AIP 1.0.0 lasting forever, I realize.  I just don't agree with that idea.

All of this is gray areas, so we will not likely agree.  I would really like to see the updates to the RFCs so that we can use them. Trust Ping I'm less concerned about if we can replace it where it is being used today (confirmation of a connection).

I'm still fine with the compromise and agree we can discuss elsewhere.

**Comment by kdenhartog** [comment]
To offer a bit of nuance to why I was arguing that AIP 1.0.0 will last forever I want to put the numbers I've been looking at out there so everyone gets where I'm coming from:

First off let's look at SSL/TLS
SSL 1.0 was never officially published, according to [Wikipedia](https://en.wikipedia.org/wiki/Transport_Layer_Security#History_and_development) and as such shows [no adoption today](https://www.ssllabs.com/ssl-pulse/) - See the protocol support graph.

However, SSL 2.0 a protocol that was published in 1995 and Deprecated in 2011 and SSL 3.0 which was published in 1996 and deprecated in 2015 still account for 7.6% of website support(refer to adoption link above). This is not to say a website only supports this, but rather that they've chosen to not turn it off even though the two protocols have been deprecated for years now.

Additionally TLS 1.0 and 1.1 will likely encounter similar issues with deprecation planned to occur this year. 

IPV4 is another perfect example of this. While IPv4 has been shown to have cracks in the foundation of number of namespaces available, but do to the fact that many people running IPv4 today can meet their needs we've yet to see broad adoption of IPv6, which has had [alpha quality code in the Linux kernel since 1996](https://en.wikipedia.org/wiki/IPv6_deployment#Major_milestones).

Furthermore, adoption of IPv6 has been incredibly slow because there's been little justifiable reason for the [complexity brought by the transition plan](https://en.wikipedia.org/wiki/IPv6_deployment#Coexistence_with_IPv4).

A third example of this is the comparison between telnet and SSH. Telnet dates back to 1969, well before my time, and was considered deprecated long ago. However, even though telnet is largely deprecated at this point, support has not been fully removed because we still have legacy systems where this is the only way to communicate with them.

So my point around this is not to say "forever" actually means forever (aka 20+ years), but it is to say that it will be around and in use longer than any of us plan to maintain and patch code for it's support. This is why I'm cautious of the requirements we are setting as absolutes and what is supported but not a requirement. It's purely because throughout the internet's history we've seen protocols linger well past their lifetimes and this will likely lead to implications on how long we have to maintain and patch EoL codebases when we'd rather be working on the current and future stuff. Hopefully this explains why I'm arguing that the problem reports should be thought through and taken to a semi mature phase (I've not formed an opinion on what that means in this case yet) before we declare AIP 1.0 set.

**Comment by esplinr** [comment]
We removed all the protocols that concerned me from AIP 1.0, except for:
* Problem Report: which I agree we should include and manage the privacy concern.
* Service Decorator: which we are still investigating and will raise a separate issue if needed.
Therefore, we have addressed the concerns that caused me to open this issue and I'll close it.
